### ISSUE TITLE
How to build optimizer

### ISSUE THREAD

**Issue by pfeatherstone** [lucidrains/x-transformers#230]
Looking at 

https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/model.py#L263

https://github.com/karpathy/minGPT/blob/37baab71b9abea1b76ab957409a1cc2fbfba8a26/mingpt/model.py#L215

https://github.com/ultralytics/ultralytics/blob/d021524e850acfa393ec25d4ecb9c4c761cca688/ultralytics/engine/trainer.py#L688

a few repositories carefully build optimizers by splitting parameters into groups, which will either experience weight decay or not. All of them agree biases of any kind don't while kernel weights from `nn.Linear`, `nn.ConvNd` do. 
This repository has many kind of parameters. 
My question is: where do they fall?

A shortlist of parameters I'm not sure about:
* `ScaleNorm.g`
* `RMSNorm.g`
* `TransformerWrapper.memory_tokens`
* `Attention.mem_k` and `Attention.mem_v`

Thank you

**Comment by pfeatherstone** [comment]
Currently i'm using:
```
def createOptimizer(model: torch.nn.Module, betas=(0.9,0.95), lr=0.001, decay=0.1):
    bn = tuple(v for k, v in nn.__dict__.items() if "Norm" in k) + (nn.Embedding,)  # normalization layers, i.e. BatchNorm2d(), and embedding
    blacklist = ["bias", "memory_tokens", 'mem_k', 'mem_v']
    decay_params   = []
    nodecay_params = []
    for module_name, module in model.named_modules():
        for param_name, param in module.named_parameters(recurse=False):
            fullname = f"{module_name}.{param_name}" if module_name else param_name
            if any(substr in fullname for substr in blacklist) or isinstance(module, bn):  
                nodecay_params.append(param)
            else:
                decay_params.append(param)

    ndecayed            = len(decay_params)
    nnodecayed          = len(nodecay_params)
    ntotal              = len(list(filter(lambda p: p.requires_grad, self.parameters())))
    assert ndecayed + nnodecayed == ntotal, f"bad split: {ndecayed} + {nnodecayed} != {ntotal}"
    optim_groups = [
        {'params': decay_params,   'weight_decay': decay},
        {'params': nodecay_params, 'weight_decay': 0.0}
    ]
    optimizer = torch.optim.AdamW(optim_groups, lr=lr, betas=betas, fused=True)
    return optimizer
```

**Comment by pfeatherstone** [comment]
Basically, i've only just started playing with optimizers and found that they have a massive influence on convergence rate and stability. Duh.

**Comment by pfeatherstone** [comment]
Can anybody think of any other layers/parameters that shouldn't decay ?

**Comment by lucidrains** [comment]
`pip install pytorch-custom-utils`

`from pytorch_custom_utils import get_adam_optimizer`

**Comment by lucidrains** [comment]
@pfeatherstone and yeah, typically you just filter out any parameters with ndims <= 1, however, i've also heard from researchers that it doesn't matter

**Comment by lucidrains** [comment]
@pfeatherstone or hop on eleutherai and consult the crowd intelligence there. everyone has their own opinions about optimizers

**Comment by lucidrains** [comment]
and before you try to consult me for this, no [don't use](https://arxiv.org/abs/2307.06440) lion or sophia or whatever

**Comment by pfeatherstone** [comment]
@lucidrains Thank you. It looks like you are doing what [nanogpt](https://github.com/karpathy/nanoGPT) is doing. I'll have a little fiddle. That does mean you are decaying normalization weights. I'll have a fiddle. Sorry if this is out of scope.

**Comment by lucidrains** [comment]
@pfeatherstone well, it isn't i'm doing what they are doing, we are both following an early practice for the original transformer training from Brain. however, whether it really matters is still up for a future research paper to decide

**Comment by lucidrains** [comment]
@pfeatherstone yea, why don't you ask the crowd at EleutherAI instead of me?
